### PR TITLE
LPS-127012 Create balloon editor react component

### PR DESCRIPTION
**Pull request by julien**
This add the `BalloonEditor` component in `frontend-editor-ckeditor-web` 
and a new `frontend-editor-ckeditor-sample-web` portlet.

As we are not replacing our current AlloyEditors (present in Blogs and Page Editor) for now (until work is completely finished), this example allows us to test the `BalloonEditor` easily in DXP. 
The idea is to evolve this portlet and render any other editors available, with a switch or any selector. 

To add this portlet, edit the main page (or either create another widget page) 
and look for **CKEditor Sample** in the widgets section. 

For now, this is what it renders (note that we still have to work in the toolbars and other stuff). 

**Preview**

### CODE REVIEW

**Comment by ambrinchaudhary** [comment]
Just started reviewing :)

**Comment by ambrinchaudhary** [comment]
I just left some more commits in top of this pull: https://github.com/julien/liferay-portal/pull/291

**Comment by julien** [comment]
@ambrinchaudhary thanks!

I thought you were going to push directly on [this branch](https://github.com/julien/liferay-portal/tree/LPS-127012), but no problem, I updated this PR with everything.

The only thing I see changed in the sample is some styles not being applied, but I wouldn't worry for now

**Review by kresimir-coko — APPROVED** [review]
Took a quick glance, not horrible 👍 
Only question I got is why are we adding a couple empty files (`init-ext.jsp` and `.lfrbuild-portal`).
I assume they're needed for the module to get hooked up with DXP properly.

**Comment by wincent** [comment]
> Only question I got is why are we adding a couple empty files (`init-ext.jsp` and `.lfrbuild-portal`).
> I assume they're needed for the module to get hooked up with DXP properly.

I think your intuition serves you well.

- **If you search for `init-ext.jsp` files in DXP, what do you find?** Hopefully, you'll see a common pattern which is that there are `init.jsp` files with setup stuff in them, and then `init-ext.jsp` files that all look like empty placeholders. 🤔 — the "ext" must stand for extension and is provided as a "blank slate" for customers to overwrite with whatever customizations that they might want, safe in the knowledge that we (Liferay, Inc) is never going to touch those files.
- **Based on the name alone and a search for other files starting with `.lfr`, what do you think this one might be for?** The fact that there are so many of these files, all empty, suggests that they are being used as a flag in some way. 🤔 — seems like the "build" here might be a flag to include the module in which it appears in the Portal build process. If you look at the other dotfiles such as `.lfrbuild-releng-ignore` you can probably come up with a good guess for what they mean too.

**Comment by ambrinchaudhary** [comment]
Removing draft... just leaving it here for a bit in case anyone wants to add his review. This is the beginning of the epic and we will continue adding stuff to this through different PRs / stories.

**Comment by jbalsas** [comment]
@julien, @ambrinchaudhary, could we spice the description up a little bit? Add instructions as to how to test the newly added editor and maybe even a screenshot for bigger impact? 😉

**Comment by ambrinchaudhary** [comment]
Just started reviewing :)

**Comment by ambrinchaudhary** [comment]
Hey @jbalsas, thanks for your comments. 

> Also, is the `ckeditor-sample-web` module meant to be:
>     Committed to master?
>     Released with DXP?

**Committed to master:** yes, we would like to have this sample portlet to easily test our component. 
**Released with DXP**: Not sure what this exaclty means, but I would say _why not_? Or maybe your point is that this module should not been deployed automatically? For which I guess we should only remove the `.lfrbuild-portal` file...

**Comment by julien** [comment]
@jbalsas , @ambrinchaudhary concerning those two questions:

> **Committed to master**

I'd say we "thought" it would be a good idea to have a sample portlet to be able to test our changes during development and we also thought it could be a good idea for QA tests, but it's not mandatory, for example we could probably try out the balloon editor in an existing module (blogs, seems like an appropriate place) if that makes more sense.

> **Released with DXP**

After thinking about it, I'd say we probably don't want that, so as @ambrinchaudhary said we should get rid of the `.lfrbuild-portal` file.

In any case @ambrinchaudhary and I should decide about the first point and update this PR (with a better description, steps to deploy/test and screenshot)

**Comment by jbalsas** [comment]
> **Released with DXP**: Not sure what this exaclty means, but I would say why not? Or maybe your point is that this module should not been deployed automatically? For which I guess we should only remove the .lfrbuild-portal file...

I meant, do we think customers should be able to use this widget and place it in pages?

Please, check the following articles to get some additional information about the release process, possible exclusions and how to enact them:
- [DXP & CE GA Build Process](https://liferay.atlassian.net/wiki/spaces/TEAMRM/pages/1385497029/DXP+CE+GA+Build+Process)
- [Feature Flags and Release Ignores in DXP - Implementation, Testing and Best Practices](https://docs.google.com/document/d/1fWbipePSQVMwKBcv6aFySws0AOTa7rfMl3Nbr3ehBpc)

**Comment by wincent** [comment]
> After thinking about it, I'd say we probably don't want that, so as @ambrinchaudhary said we should get rid of the `.lfrbuild-portal` file.

I don't think that's how it works:

- `.lfrbuild-portal` says to _build_ the module during `ant all`; and:
- `.lfrbuild-releng-ignore` says to exclude it from release.

At least, that's my understanding.

**Comment by julien** [comment]
@wincent on top of what you said, I understood that having a `.lfrbuild-portal` file means having to upload jars to the nexus server; and having a `.lfrbuild-releng-ignore` means that the module will not be uploaded to maven ... 

I'll keep on reading because I still don't understand all the implications and would rather not include something unless it's strictly required.

**Comment by wincent** [comment]
> @wincent on top of what you said, I understood that having a `.lfrbuild-portal` file means having to upload jars to the nexus server; and having a `.lfrbuild-releng-ignore` means that the module will not be uploaded to maven ...

You speak to me as if I knew what "nexus" was. 😂 My understanding is that it is "a server, or something, that holds files, or something". Approximately. I don't know what the ramifications are, but my uninformed guess is that being uploaded to Liferay's Nexus servers doesn't really have any implications for customers at all.

**Comment by jbalsas** [comment]
>  I understood that having a .lfrbuild-portal file means having to upload jars to the nexus server; and having a .lfrbuild-releng-ignore means that the module will not be uploaded to maven ...

Having a `.lfrbuild-portal` marker is what makes the module compile and deploy in a normal `ant all` run. It's what makes most modules available by default when you build portal every morning.

A module lacking an `.lfrbuild-portal`: 
- Won't be compiled by default and deployed to your Liferay Portal instance
- Can be deployed manually
- Won't be published to Nexus (our maven repository) and thus can't be consumed as a dependency in your `build.gradle` or `build.pom`
- Won't be included in the released bundles (not available to customers)

A module with a `.lfrbuild-portal` and a `.lfrbuild-releng-ignore`:
- Will be compiled by default and deployed to your Liferay Portal instance
- Won't be published to Nexus (our maven repository) and thus can't be consumed as a dependency in your `build.gradle` or `build.pom`
- Won't be included in the released bundles (not available to customers)

This is my understanding 👆 . I actually [asked @jpince for clarification in Confluence](https://liferay.atlassian.net/wiki/spaces/TEAMRM/pages/1385497029/DXP+CE+GA+Build+Process#Release-Ignores)

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js
Comment (jbalsas) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:36]: Just a thought... couldn't we use `ClassicEditor` as a base here? It feels like we've added a lot of fixes in there that we might need to replicate here every time and that could be avoided if we already use a better base.

Just wondering why the decision of building this from the ground up and not on top of it.

Comment (julien) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:36]: I mainly avoided it, because `BalloonEditor` won't support as many features as `ClassicEditor`, for example [toolbarSets](https://github.com/liferay/liferay-portal/blob/9c07f7bca2bbde25a8cbcf2c84b6b311fb60e0f7/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/ClassicEditor.js#L38) which probably won't be defined the same way (considering that "BalloonToolbars" aren't exactly the same as regular toolbars in CKEditor), I wasn't sure if we wanted to maintain [this global API](https://github.com/liferay/liferay-portal/blob/9c07f7bca2bbde25a8cbcf2c84b6b311fb60e0f7/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/ClassicEditor.js#L101-L108) either and I wanted to avoid [this resize handler](https://github.com/liferay/liferay-portal/blob/9c07f7bca2bbde25a8cbcf2c84b6b311fb60e0f7/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/ClassicEditor.js#L101-L108)  - but I'll try building `BalloonEditor` on top of `ClassicEditor` and see how it goes.

**Comment by julien** [comment]
@jbalsas thanks for the clarification, based on that I think we'll go with  

> A module with a .lfrbuild-portal and a .lfrbuild-releng-ignore:
> - Will be compiled by default and deployed to your Liferay Portal instance
> - Won't be published to Nexus (our maven repository) and thus can't be consumed as a dependency in your build.gradle or build.pom
> - Won't be included in the released bundles (not available to customers)

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js
Comment (jbalsas) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:36]: No, that's okay, sounds good to me. Just wanted to make sure we had a good reason for this. We can always create a common base if we see a lot of duplication between the editors.

Let's just add the release-ignore markers and push this forward if ready!

Comment (julien) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:36]: :+1:

**Comment by julien** [comment]
ci:test:sf

**Comment by ambrinchaudhary** [comment]
ci:test:relevant

**Review by ambrinchaudhary — APPROVED** [review]

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
Going to rebase with master and re-launch ci tests.

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
Looking at the test report, I see this 

```
format-source-files:
     [java] java.lang.Exception: Found 1 formatting issues:
     [java] 1: ./build.properties expected:<...al,\
     [java]     #    apps/s[]taging,\
     [java]     #    ap...> but was:<...al,\
     [java]     #    apps/s[ocial-bookmarks,\
     [java]     #    apps/s]taging,\
     [java]     #    ap...>
     [java] 
     [java] 	at com.liferay.source.formatter.SourceFormatter.format(SourceFormatter.java:456)
     [java] 	at com.liferay.source.formatter.SourceFormatter.main(SourceFormatter.java:299)
```

And when I run `ant format-source-files` from the `portal-impl` directory locally I get different results, so I'm a bit lost here

**Comment by julien** [comment]
I mentioned this issue on [slack](https://liferay.slack.com/archives/C19PWQ74J/p1617868824095800), and a fix is on it's way (:partying_face:), I don't think it's a problem if we wait until it's merged to forward this.

**Comment by ambrinchaudhary** [comment]
Just started reviewing :)

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
ci:forward

**Pull request closed without merging**
